packages/adapter-components/src/filters/service_url.ts

```diff
@@ -45,38 +72,73 @@ export const addUrlToInstance = (instance: InstanceElement, baseUrl: string, api
+  Action extends string = ActionName,
+>(
+  baseUrl: string,
+) => AdapterFilterCreator<TContext, TResult, TAdditional, ClientOptions, PaginationOptions, Action> =
+  baseUrl =>
+  ({ definitions }) => {
+    if (definitions.fetch === undefined) {
+      log.warn('No fetch definitions were found, skipping service url filter')
+      return () => ({})
+    }
+    const { instances } = definitions.fetch
+    const defQuery = queryWithDefault(instances)
+    return {
+      name: 'serviceUrlFilter',
+      onFetch: async (elements: Element[]) => {
+        elements.filter(isInstanceElement).forEach(instance => {
+          addUrlToInstance(instance, baseUrl, defQuery.query(instance.elemID.typeName))
+        })
+      },
+      onDeploy: async (changes: Change<InstanceElement>[]) => {
+        const relevantChanges = changes.filter(isInstanceChange).filter(isAdditionChange)
+        relevantChanges.map(getChangeData).forEach(instance => {
+          addUrlToInstance(instance, baseUrl, defQuery.query(instance.elemID.typeName))
+        })
+      },
+    }
+  }
+// TODO deprecate when upgrading to new definitions SALTO-5538
+export const serviceUrlFilterCreatorDeprecated: <
   TClient,
   TContext extends { apiDefinitions: AdapterApiConfig },
   TResult extends void | filter.FilterResult = void,
 >(
   baseUrl: string,
   additionalApiDefinitions?: AdapterApiConfig,
-) => FilterCreator<TClient, TContext, TResult> =
-  (baseUrl, additionalApiDefinitions) =>
-  ({ config }) => ({
-    name: 'serviceUrlFilter',
-    onFetch: async (elements: Element[]) => {
-      elements.filter(isInstanceElement).forEach(instance => {
-        const apiDefinitions =
-          additionalApiDefinitions?.types[instance.elemID.typeName] !== undefined
-            ? additionalApiDefinitions
-            : config.apiDefinitions
-        addUrlToInstance(instance, baseUrl, apiDefinitions)
-      })
-    },
-    onDeploy: async (changes: Change<InstanceElement>[]) => {
-      const relevantChanges = changes.filter(isInstanceChange).filter(isAdditionChange)
-      relevantChanges.map(getChangeData).forEach(instance => {
-        const apiDefinitions =
-          additionalApiDefinitions?.types[instance.elemID.typeName] !== undefined
-            ? additionalApiDefinitions
-            : config.apiDefinitions
-        addUrlToInstance(instance, baseUrl, apiDefinitions)
-      })
+) => FilterCreator<TClient, TContext, TResult> = (baseUrl, additionalApiDefinitions) => args => {
+  const { config } = args
+  const typeNamesSet = new Set([
+    ...Object.keys(config.apiDefinitions.types),
+    ...Object.keys(additionalApiDefinitions?.types ?? {}),
+  ])
+  const customizations = Object.fromEntries(
+    Array.from(typeNamesSet).map(typeName => [
+      typeName,
+      configDefToInstanceFetchApiDefinitionsForServiceUrl(
+        config.apiDefinitions.types[typeName],
+        additionalApiDefinitions?.types[typeName],
+      ),
+    ]),
+  )
```

Comment:
I think this can be simpler - e.g.
```suggestion
  const customizations = _({ ...config.apiDefinitions.types, ...additionalApiDefinitions?.types,})
    .mapValues(configDefToInstanceFetchApiDefinitionsForServiceUrl)
    .pickBy(lowerdashValues.isDefined)
    .value()
```
and we can have `configDefToInstanceFetchApiDefinitionsForServiceUrl` accept a single arg. wdyt?
(but maybe I'm missing something)

Reply:
You're not missing anything, just using a lot of magic that I wasn't familiar with :)